src/main/java/net/dries007/tfc/client/TFCColors.java

```diff
@@ -125,25 +137,61 @@ public static int getWaterFogColor(CommonLevelAccessor level, Biome biome, Block
         return TFCBiomes.hasExtension(level, biome) ? getClimateColor(WATER_FOG_COLORS_CACHE, pos) : biome.getWaterFogColor();
     }
 
-    public static int getSeasonalFoliageColor(@Nullable BlockPos pos, int tintIndex)
+    public static int getSeasonalFoliageColor(@Nullable BlockPos pos, int tintIndex, BlockState state)
     {
-        if (pos != null && tintIndex == 0)
+        final Level level = ClientHelpers.getLevel();
+        if (level != null && pos != null)
         {
-            switch (getAdjustedNoisySeason(pos))
-            {
-                case SPRING:
-                case SUMMER:
-                    return getClimateColor(FOLIAGE_COLORS_CACHE, pos);
-                case FALL:
-                    int index = Helpers.hash(91273491823412341L, pos);
-                    return FOLIAGE_FALL_COLORS_CACHE[index & COLORMAP_MASK];
-                case WINTER:
-                    return getClimateColor(FOLIAGE_WINTER_COLORS_CACHE, pos);
-            }
+            return getSeasonalFoliageColor(pos, tintIndex, level, state);
+        }
+        return -1;
+    }
+
+    public static int getSeasonalFoliageColor(BlockPos pos, int tintIndex, LevelAccessor level, BlockState state)
+    {
+        if (tintIndex == 0)
+        {
+            return getSeasonalFoliageColor(pos, level, state);
         }
         return -1;
     }
 
+    /**
+     * Gets a color based on average temperature and time of year. Autumn occurs at different times of the year at height-adjusted average temperatures from the poles to 14c
+     */
+    private static int getSeasonalFoliageColor(BlockPos pos, LevelAccessor level, BlockState state)
+    {
+        ChunkData data = ChunkData.get(level, pos);
+        float temp = data.getAdjustedAverageTempByElevation(pos);
+        float timeOfYear = Calendars.CLIENT.getCalendarFractionOfYear();
+        final float tempClamped = temp > 14f ? 14f : Math.max(temp, -20f);
+
+        final float cubedTerm = 1.5f * (float) Math.pow(tempClamped + 3f, 3f) / 4913f;
+        final float squaredTerm = 0.5f * (float) Math.pow(tempClamped + 3f, 2f) / 289f;
+        final float autumnEnd = (cubedTerm - squaredTerm + 10.5f) / 12f;
+        final float autumnStart = temp > 14f ? autumnEnd : (cubedTerm + squaredTerm + 8.5f) / 12f;
+        final float springStart = 1f - autumnEnd;
+
+        if (timeOfYear > autumnEnd)
+        {
+            return getClimateColor(FOLIAGE_WINTER_COLORS_CACHE, pos);
+        }
+        else if (timeOfYear > autumnStart)
+        {
+            final Block block = state.getBlock();
+            final int autumnIndex = block instanceof TFCLeavesBlock ? ((TFCLeavesBlock) block).getAutumnIndex() : 127;
```

Comment:
try `instanceof TFCLeavesBlock block`

I wonder if this should be on the `ILeavesBlock` interface so that for example I can implement the interface on a modded leaf block (for dynamic trees for example)

Reply:
See my other comments, this can avoid being on the block, the interface, or the state, and just part of registration which is as easy to implement for modded as possible.